extensions/wikia/Captcha/Captcha.class.php

```diff
@@ -0,0 +1,625 @@
+			if ( $this->isIPWhitelisted() )
+				return true;
+
+			$this->trigger = "new account '" . $u->getName() . "'";
+			if ( !$this->captcha->passCaptcha() ) {
+				$message = wfMsg( 'captcha-createaccount-fail' );
+				return false;
+			}
+		}
+		return true;
+	}
+
+	/**
+	 * Hook for user login form submissions.
+	 *
+	 * @param \User $u
+	 * @param $pass
+	 * @param $retval
+	 *
+	 * @return bool true to continue, false to abort user creation
+	 */
+	public function confirmUserLogin( $u, $pass, &$retval ) {
+		if ( $this->isBadLoginTriggered() ) {
+			if ( $this->isIPWhitelisted() )
+				return true;
+
+			$this->trigger = "post-badlogin login '" . $u->getName() . "'";
+			if ( !$this->captcha->passCaptcha() ) {
+				// Emulate a bad-password return to confuse the shit out of attackers
+				$retval = \LoginForm::WRONG_PASS;
+				return false;
+			}
+		}
+		return true;
+	}
+
+	/**
+	 * Check the captcha on Special:EmailUser
+	 *
+	 * @param \MailAddress $from
+	 * @param \MailAddress $to
+	 * @param string $subject
+	 * @param string $text
+	 * @param string $error reference
+	 *
+	 * @return Bool true to continue saving, false to abort and show a captcha form
+	 */
+	public function confirmEmailUser( $from, $to, $subject, $text, &$error ) {
+		$wg = \F::app()->wg;
+		if ( $wg->CaptchaTriggers['sendemail'] ) {
+			if ( $wg->User->isAllowed( 'skipcaptcha' ) ) {
+				$this->log( "user group allows skipping captcha on email sending\n" );
+				return true;
+			}
+			if ( $this->isIPWhitelisted() ) {
+				return true;
+			}
+
+			if ( defined( 'MW_API' ) ) {
+				# API mode
+				# Asking for captchas in the API is really silly
+				$error = wfMsg( 'captcha-disabledinapi' );
+				return false;
+			}
+			$this->trigger = "{$wg->User->getName()} sending email";
+			if ( !$this->captcha->passCaptcha() ) {
+				$error = wfMsg( 'captcha-sendemail-fail' );
+				return false;
+			}
+		}
+		return true;
+	}
+
+	/**
+	 * Log the status and any triggering info for debugging or statistics
+	 *
+	 * @param string $message
+	 */
+	public function log( $message ) {
+		$log = WikiaLogger::instance();
+		$log->info( 'Captcha: ' . $message, [
+			'action' => $this->action,
+			'detail' => $this->trigger,
+		] );
+	}
+
+	/**
+	 * Retrieve the current version of the page or section being edited...
+	 *
+	 * @param \EditPage $editPage
+	 * @param string $section
+	 *
+	 * @return string
+	 */
+	private function loadText( $editPage, $section ) {
+		$rev = \Revision::newFromTitle( $editPage->mTitle );
+		if ( is_null( $rev ) ) {
+			return "";
+		} else {
+			$text = $rev->getText();
+			if ( $section != '' ) {
+				return \F::app()->wg->Parser->getSection( $text, $section );
+			} else {
+				return $text;
+			}
+		}
+	}
+
+	/**
+	 * Extract a list of all recognized HTTP links in the text.
+	 *
+	 * @param \EditPage $editPage
+	 * @param string $text
+	 *
+	 * @return array of strings
+	 */
+	public function findLinks( &$editPage, $text ) {
```

Comment:
Does this method need to be public? Might be worth a pass to see if all the public methods need to be. This seems like it's only being used in this class. Probably not a big deal but could cut down on ambiguity.